pkg/sql/sem/builtins/builtins.go

```diff
@@ -2304,8 +2305,26 @@ var jsonTypeOfImpl = tree.Builtin{
 	Types:      tree.ArgTypes{{"val", types.JSON}},
 	ReturnType: tree.FixedReturnType(types.String),
 	Fn: func(_ *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
-		json := tree.MustBeDJSON(args[0])
-		return tree.NewDString(json.TypeAsText()), nil
+		djson := tree.MustBeDJSON(args[0])
+		var typeOf string
+		t := djson.JSON.Type()
+		switch t {
+		case json.NullJSONType:
+			typeOf = "null"
```

Comment:
With these, it's a small thing, but it might make more sense to just allocate these once ahead of time:

```go
var jsonNullDString = tree.NewDString("null")
...etc
```
above and then they can be shared among all calls to this function, and don't have to reallocate on each call, does that make sense? And then each arm of the switch can be
```go
return jsonNullDString, nil
...etc
```

And then in the case of one we don't expect, while it shouldn't happen, it's probably a little safer to just return an error rather than panicking, something like
```go
pgerror.NewErrorf(pgerror.CodeInternalError, "unexpected JSON type %d", t)
```